ledger/ledger.go

```diff
@@ -40,73 +45,145 @@ type Ledger interface {
 	RootHash() string
 	// GetPreviousValue executes a get against a previous version of the ledger, using that version's root hash.
 	GetPreviousValue(previousRootHash, key string) (result string, err error)
+	// EraseRootHash re-claims any memory used by this version of history, preserving bits shared with other versions.
+	EraseRootHash(rootHash string) error
+	// Stats gives basic storage info regarding the ledger's underlying cache
+	Stats() cache.Stats
+	// GetAll returns the entire state of the ledger
+	GetAll() (map[string]string, error)
+	// GetAllPrevious returns the entire state of the ledger at an arbitrary version
+	GetAllPrevious(string) (map[string]string, error)
 }
 
 type smtLedger struct {
-	tree *smt
+	tree    *smt
+	history *history
+	// this map will grow forever.  how do we limit it?
+	keyCache byteCache
 }
 
 // Make returns a Ledger which will retain previous nodes after they are deleted.
-func Make(retention time.Duration) Ledger {
-	return smtLedger{tree: newSMT(hasher, nil, retention)}
+// the retention parameter has been removed in favor of EraseRootHash, but is left
+// here for backwards compatibility
+func Make(_ time.Duration) Ledger {
+	return &smtLedger{
+		tree:    newSMT(hasher, nil),
+		history: newHistory(),
+		// keyCache should have ~512kB memory max, each entry is 128 bits = 2^23/2^7 = 2^16
+		keyCache: byteCache{cache: cache.NewLRU(forever, time.Minute, math.MaxUint16)},
+	}
+}
+
+func (s *smtLedger) EraseRootHash(rootHash string) error {
+	// occurrences is a list of every time in (underased) history when this hash has been observed
+	occurrences := s.history.Get(rootHash)
+	if len(occurrences) == 0 {
+		return fmt.Errorf("rootHash %s is not present in ledger history", rootHash)
+	}
+	var adjacentRoots [][]byte
+	for _, o := range occurrences {
+		adjacentRoots = append(adjacentRoots, o.Prev().Value.([]byte), o.Next().Value.([]byte))
+	}
+	err := s.tree.Erase(occurrences[0].Value.([]byte), adjacentRoots)
+	if err != nil {
+		return err
+	}
+	for _, o := range occurrences {
+		s.history.Remove(o)
+	}
+	s.history.lock.Lock()
+	defer s.history.lock.Unlock()
+	delete(s.history.index, rootHash)
```

Comment:
nit: this seems like it should be a method on history rather than directly accessed